### ISSUE TITLE
Incorrect CSV column analyzing

### ISSUE THREAD

**Issue by giangnvt** [sqlitebrowser/sqlitebrowser#1032]
### Details for the issue
It seems there is a bug when parsing a CSV with some escaped double quote.
For the below CSV data, it parse incorrectly, for the 1st row, it return 6 columns, and 7 columsn for 2nd row (correct is 6), that causes the entire import to fail.

.................................................................
"串","20018","shish kebab","456590","95","naka naka , \"who's there?\" shish kebab .... \"shish kebab who?\""
"串","20018","shish kebab","468238","28","The actual definition of this kanji is \"#a skewer#,\" which I believe is much more helpful as a primitive. Just think, two *mouths* #skewer#ed on same *stick*."
.................................................................

#### I'm opening this issue because:

- [ ] DB4S is crashing
- [ ] DB4S has a bug
- [ ] DB4S needs a feature
- [x] DB4S has another problem

#### I'm using DB4S on:

- [x] Windows: ( _version:_ ___ )
- [ ] Linux: ( _distro:_ ___ )
- [ ] Mac OS: ( _version:_ ___ )
- [ ] Other: ___

#### I'm using DB4S version:

- [ ] 3.9.1
- [x] 3.9.0
- [ ] Other: ___

#### I have also:

- [ ] Tried out the latest nightly version: https://github.com/sqlitebrowser/sqlitebrowser#nightly-builds
- [ ] Searched for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by justinclift** [comment]
Looking through this, and the [CSV spec](https://www.rfc-editor.org/rfc/rfc4180.txt) (section 2.7), it seems like the example you've given has the wrong characters for escaping quotes.

Is the program you're creating the CSV with able to change the quote escape characters? :smile:

**Comment by chrisjlocke** [comment]
I don't think the hash is escaping the quotations though - in the second example, the hash is before a comma, not after the quotation (whiile in the first example, its after a quotation).

The proper delimitation should be double quotes, so the string should be:

"串","20018","shish kebab","468238","28","The actual definition of this kanji is ""#a skewer#,"" which I believe is much more helpful as a primitive. Just think, two mouths #skewer#ed on same stick."

But I might be wrong..... ;)

**Comment by justinclift** [comment]
Ahhh, I was looking at the actual example data (via the edit button), not what GitHub's showing:

```
"串","20018","shish kebab","456590","95","naka naka , \"who's there?\" shish kebab .... \"shish kebab who?\""
"串","20018","shish kebab","468238","28","The actual definition of this kanji is \"#a skewer#,\" which I believe is much more helpful as a primitive. Just think, two *mouths* #skewer#ed on same *stick*."
```

**Comment by chrisjlocke** [comment]
Correctly delimited incorrectly delimited comma separated values .... CDIDCSV?  Thank God we're not working with regular expressions....

Slightly off topic, but I posted this the other day:
![image](https://cloud.githubusercontent.com/assets/7872218/26721530/15e55b2c-4784-11e7-9082-14347dac459a.png)

**Comment by MKleusberg** [comment]
Yeah, the problem here is the escaping of the quote chars. It needs to be ```"something""something"``` instead of ```"something\"something"```. That said, we're not handling the wrong escaping very gracefully. For starters the quote chars in the middle of the field are lost which is definitely wrong. But we're also stumbling upon the comma after the second pound char in the second line. This is causing the extra column. But honestly I'm not sure how to improve the behaviour here. The quote before the first pound char ends the quoted field and with the next comma a new column will be added.

**Comment by justinclift** [comment]
@giangnvt ping? :smile:

**Comment by giangnvt** [comment]
I don't know your being used CSV parsing policy, but on my opinion, a general CSV should allow a customizable escape character, instead of a fixed double quote (I remember that's the way OpenCSV works, but not quite sure).

> This is causing the extra column. But honestly I'm not sure how to improve the behaviour here. The quote before the first pound char ends the quoted field and with the next comma a new column will be added.
If the backslash (\) is accepted as escape character, it can simply skip both the double quote before and after the pound pair, or I'm misunderstanding something?

**Comment by justinclift** [comment]
@MKleusberg Would the newly rewritten CSV importing code be able to handle `""` escaping better now?
